components/cloud/src/blob.rs

```diff
@@ -0,0 +1,191 @@
+use futures_io::AsyncRead;
+pub use kvproto::backup::CloudDynamic;
+use std::io;
+use std::marker::Unpin;
+
+/// An abstraction for blob storage.
+/// Currently the same as ExternalStorage
+pub trait BlobStorage: 'static + Send + Sync {
+    fn name(&self) -> &'static str;
+
+    fn url(&self) -> io::Result<url::Url>;
+
+    /// Write all contents of the read to the given path.
+    fn write(
+        &self,
+        name: &str,
+        reader: Box<dyn AsyncRead + Send + Unpin>,
+        content_length: u64,
+    ) -> io::Result<()>;
+
+    /// Read all contents of the given path.
+    fn read(&self, name: &str) -> Box<dyn AsyncRead + Unpin + '_>;
+}
+
+impl BlobStorage for Box<dyn BlobStorage> {
+    fn name(&self) -> &'static str {
+        (**self).name()
+    }
+
+    fn url(&self) -> io::Result<url::Url> {
+        (**self).url()
+    }
+
+    fn write(
+        &self,
+        name: &str,
+        reader: Box<dyn AsyncRead + Send + Unpin>,
+        content_length: u64,
+    ) -> io::Result<()> {
+        (**self).write(name, reader, content_length)
+    }
+
+    fn read(&self, name: &str) -> Box<dyn AsyncRead + Unpin + '_> {
+        (**self).read(name)
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct StringNonEmpty(String);
+impl StringNonEmpty {
+    pub fn opt(s: String) -> Option<Self> {
+        if s.is_empty() {
+            None
+        } else {
+            Some(Self(s))
+        }
+    }
+
+    pub fn opt2(s1: String, s2: String) -> Option<Self> {
+        Self::opt(s1).or_else(|| Self::opt(s2))
+    }
+
+    pub fn required_field(s: String, field: &str) -> io::Result<Self> {
+        Self::required_msg(s, &format!("field {}", field))
+    }
+
+    pub fn required_field2(s1: String, s2: String, field: &str) -> io::Result<Self> {
+        match Self::opt2(s1, s2) {
+            Some(sne) => Ok(sne),
+            None => Err(Self::error_required(&format!("field {}", field))),
+        }
+    }
+
+    pub fn required_msg(s: String, msg: &str) -> io::Result<Self> {
+        if !s.is_empty() {
+            Ok(Self(s))
+        } else {
+            Err(Self::error_required(&format!("Empty {}", msg)))
+        }
+    }
+
+    pub fn error_required(msg: &str) -> io::Error {
+        io::Error::new(io::ErrorKind::InvalidInput, msg)
+    }
+
+    pub fn required(s: String) -> io::Result<Self> {
+        Self::required_msg(s, "string")
+    }
+
+    pub fn stat(s: &'static str) -> Self {
+        Self::required_msg(s.to_owned(), "static str").unwrap()
+    }
+}
+
+impl std::ops::Deref for StringNonEmpty {
+    type Target = String;
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl std::fmt::Display for StringNonEmpty {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct BucketConf {
```

Comment:
Can this be
```
pub struct BucketConf(backup::CloudDynamic);
```

Reply:
I like having separate types for API and for domain modeling. protobuf types are very difficult to work with because we should be supporting both protobuf and prost and everything about their usage is different. There is an overhead to copying data around, but it doesn't matter for this metadata.